[chore] add codeowners and pub flow

carsonfarmer opened the pull request:
## Summary

Adds a code owners file, plus creates a deploy action and flow for automatically publishing to npm.

## Details

Fixes #61 
Fixes #140 

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

joewagner reviewed (dismissed):
Looks good, but github highlighting says the CODEOWNERS file is invalid because of line 8.  Specifically `{js,ts}` is not a valid owner.  Not sure if that matters at all?

sanderpick commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,49 @@
+name: Publish
```
nice! looks like this happens for every merge and publishes to `next` version? I think we also want a flow that kicks off from a tag and uses that tag number as new NPM version number, right?

carsonfarmer replied:
This will only do the release if the version number has been incremented from what is already on npmjs. It will always default to next. It is then up to us to (when appropriate) tag a given release as "latest" manually, when we do a release. We _could_ actually automate this secondary step based on a release action...

carsonfarmer replied:
I _might_ be able to set this up so that if we tag a commit with something like "latest" it'll do that rather than next, but this next staging flow felt like a nice catch for potentially catching breaking updates without going straight to "latest".

carsonfarmer replied:
Basically the idea is, automated builds _don't_ go to "latest". To move a release to "latest", it moves through "next", and then "latest".

sanderpick replied:
imo, the tag-based flow (like we have in go-threads and elsewhere) is nice because you don't have to mess with the NPM file at all. You just create release in GH, which triggers an action that updates the NPM file with the tag name, publishes to NPM. That means just one click for the dev.

With the current setup, do you need a manual interaction on the NPM page to mark next as latest?

carsonfarmer replied:
Yes, with the current flow, you need to manually mark a given "next" tagged release as "latest". It is slightly more complicated here than in the go case (which was an issue with tools like lerna) because we actually have to update the package.json file and commit that if we do fully off tagging a release. However, we _could_ come up with a tagging mechanism so that you simply bump the version and use either "next" or "latest" in the commit message, similar to how we do it with go?

sanderpick replied:
Oh, I'm referring to the gRPC NPM modules we publish from go-threads, not the Go modules: https://github.com/textileio/go-threads/blob/master/.github/workflows/publish-grpc-libs.yml#L63. Basically, we leave the version 0.0.0 or whatever in the actual code... everything happens in CD, which keeps NPM module versions exact matched to GH releases. It works with `rc` / `dev` releases as well, since you can mark a tag like `1.2.3-rc.4` as "pre release". 

I think what you have is workable, so long as we remember to create GH releases with change logs. What we did in go-threads gRPC libs is one click (create release) vs. four (create PR with NPM version bump, merge it, change next to latest, create GH release to match).

carsonfarmer replied:
Hmmm, I'm not a fan of the 0.0.0 version of a project with its own repo. But some hybrid version of these two approaches could work for sure. We _can_ automate next versus latest, I just wanted to have that "upgrade" be an explicit action on our part. There are always ways to commit the updated package.json, which I think @asutula does in some other repos. It just requires a bot with permissions. Anyway, I'm happy to incorporate some additional automations here to reduce boilerplate work, but I'd like it to match best practices with other JS repos as well. So will come up with some tweaks.

carsonfarmer replied:
Oh I think I have a nice plan for this...

sanderpick replied:
sounds good. what you have here is great... actually i need to publish a new version of this to get the string helpers, so let's put it to work!

carsonfarmer commented:
> Looks good, but github highlighting says the CODEOWNERS file is invalid because of line 8. Specifically `{js,ts}` is not a valid owner. Not sure if that matters at all?

Thanks for pointing that out. Fixed and also added @sanderpick

joewagner reviewed (approved):
LGTM!

carsonfarmer commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,49 @@
+        uses: JS-DevTools/npm-publish@v1
+        with:
+          token: ${{ secrets.NPM_AUTH_TOKEN }}
+          tag: next
```
We always push "automated" releases to "next" tag. It is then up to project owners to upgrade a given release to "latest" manually. This could be optimized further if we don't like this flow.

carsonfarmer commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,49 @@
+          token: ${{ secrets.NPM_AUTH_TOKEN }}
+          tag: next
+          access: public
+          check-version: true
```
This action will only _actually_ publish a new version to npm iif the version in the incoming package.json is greater than the version already on npm. Note that dev versions with lager prefixes _do_ satisfy this constraint if we need to use them.
